proposals/fibers/Explainer.md

```diff
@@ -0,0 +1,874 @@
+
+Notice that only the newly created fiber has access to the reference that identifies the new fiber. If it is desired that the parent has access to this the code generator would typically insert a `fiber.suspend` instruction at the start of the function and pass out the `fiber` in an appropriate event description.
+
+#### `fiber.suspend` Suspend an active fiber
+
+The `fiber.suspend` instruction takes a fiber as an argument and suspends the fiber. The identified fiber must either be the `active` fiber, or a resume ancestor of the active fiber. 
+
+The _root_ ancestor fiber does not have an explicit identifier; and so it may not be suspended.
+
+The fiber that is suspended is marked `suspended`, and the the immediate resume parent of that fiber becomes the active fiber.
+
+`fiber.suspend` has two operands: the identity of the fiber being suspended and a description of the event it is signaling: the `event` tag and any arguments to the event. The event operands must be on the argument stack.
+
+Execution of a fiber that issues the `fiber.suspend` instruction will not continue until another fiber issues the appropriate `fiber.resume` instruction. In which case, execution of the suspending fiber will continue with the `event` block whose tag was used to resume it. In effect, this means that the instructions immediately following the `fiber.suspend` are _unreachable_.
+
+#### `fiber.resume` Resume a suspended fiber
+
+The `fiber.resume` instruction takes a fiber as argument, together with an `event` description&mdash;consisting of an event tag and appropriate values, and resumes its execution.
+
+The `fiber.resume` instruction takes a `suspended` fiber, together with any descendant fibers that were suspended along with it, and resumes its execution. The event is used to encode how the resumed fiber should react: for example, whether the fiber's requested information is available, or whether the fiber should enter into cancelation mode.
+
+A `fiber.resume` instruction may continue in one of two ways: if the resumed fiber returned normally, then execution continues with the next instruction. The values on the value stack will match the return signature of the fiber's function&mdash;and will match the signature of the `fiber` reference that was resumed.
+
+If the resumed fiber suspended itself, then the event tag associated with that `fiber.suspend` instruction is used to determine which of the available `event` blocks should be entered as part of the switch. The 'nearest' `event` block whose tag is equal to the supplied event is entered. If there is no appropriate `event` block in the execution scope of the fiber being resumed, then the engine _traps_.
+
+#### `fiber.switch` Switch to a different fiber
```

Comment:
So this is like a `return_call` except instead of returning it is suspending and instead of calling it is resuming?

Reply:
fiber.switch is similar to a combination of fiber.suspend + fiber.resume. However, the event involved is never inspected by the fiber's resume parent.

Reply:
Just like how the operands to `return_call` are never inspected by the caller of the return-calling function.

Reply:
return_call has a requirement that the return types of the called function match the return types of the calling function.
However, in the case of a fiber switching between fibers, the types of the event need not be the same as the scheduler's event (or, more accurately, the scheduler does not need to know how to handle the switch event itself.
This can be important for symmetric coroutining and for implementation channels a la go-lang. When a channel variable gets bound, you can directly wake up the partner goroutine if it has previously blocked on the channel. But channels themselves are not necessarily known to the scheduler.

Reply:
Having said all that, IMO, it is not obvious that we want this. There is another 'action at a distance' thing here: the target fiber may throw an exception. And since exceptions *are* propagated (ugh) that can cause the scheduler to be asked to handle an exception that it was not expecting (or worse, it can handle it incorrectly). The true cost of this risk in not easily quantifiable. IMO, we should not propagate exceptions; but that ship has sailed.